iptables: Add rules runtime reconciliation

pippolo84 opened the pull request:
As an attempt to make the agent sensitive to runtime changes, the iptables manager reinstalls all the iptables rules, including the ones related to the proxies, when the datapath is explicitly reinitialized.

This approach does not consider the updates to the external facing network devices, that are read only once at the module startup through the daemon configuration options.

This PR adds a reconciler to update the rules following any updates to the external facing network devices or to the local node.  The reconciler listens for those updates and independently rewrites all the rules accordingly, keeping into account proxy rules and no track pod rules too.

This reconciler is not based on the stateDB generic one but is designed from scratch for this specific case. This is needed since the order in which the iptables rules are installed in the chains matters, and the stateDB generic reconciler does not care about ordering.

To migrate to this new asynchronous reconciliation approach, the manager is reworked so that the methods exposed do not immediately reconcile the rules, but collect the desired state and ask the underlying reconciler to rewrite the rules accordingly.

With this new iptables manager implementation the rules reinstallation request from the datapath reinitilization is not needed anymore, hence it is removed.

For more details, please refer to the individual commit messages.

**Note to the reviewers**: please review per commit.

Related: https://github.com/cilium/cilium/pull/30024

Depends on ~https://github.com/cilium/cilium/pull/31068~, ~https://github.com/cilium/cilium/pull/31099~ and ~https://github.com/cilium/cilium/pull/31368~

pippolo84 committed:
iptables: Add rules runtime reconciliation

As an attempt to make the agent sensitive to runtime changes, the
iptables manager reinstalls all the iptables rules, including the ones
related to the proxies, when the datapath is explicitly reinitialized.

This approach does not consider the updates to the external facing
network devices, that are read only once at the module startup through
the daemon configuration options.

This commit adds a reconciler to update the rules following any updates
to the external facing network devices or to the local node.  The
reconciler listens for those updates and independently rewrites all the
rules accordingly, keeping into account proxy rules and no track pod
rules too.

This reconciler is not based on the stateDB generic one but is designed
from scratch for this specific case. This is needed since the order in
which the iptables rules are installed in the chains matters, and the
stateDB generic reconciler does not care about ordering.

To migrate to this new asynchronous reconciliation approach, the manager
is reworked so that the methods exposed do not immediately reconcile the
rules, but collect the desired state and ask the underlying reconciler
to rewrite the rules accordingly.

To rate limit the calls to the iptables command, a ticker is used to
decide when to reconcile.  This also has the advantage of coalescing a
burst of changes into a single call to the iptables/ip6tables commands.
Since a rewrite operation can be pretty expensive in terms of time, this
results in an overall lower latency for those burst updates.

With this new iptables manager implementation the rules reinstallation
request from the datapath reinitilization is not needed anymore, hence
it is removed.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Add ip{v4,v6} native routing CIDR

Manage ipv4 and ipv6 native routing CIDRs in the local node struct. This
allows the iptables reconciler to subscribe to the local node update
and:

- get the native routing CIDRs set at startup as a config option
- get the updated v4 native routing CIDRs in case of auto detection
  (when IPAM is one of CRD, ENI, Azure or Alibaba) at runtime

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add proxy rules incremental reconciliation

A full iptables rules rewrite is pretty expensive, as all the chains
must be rewritten by the iptables/ip6tables command.

To speed up the reconciliation in case of a proxy rule update, the
reconciler first tries to synchronously update only the proxy related
rules, falling back to a full asynchronous rules reconciliation only in
case of a failure.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add no track rules incremental reconciliation

As already done for the proxy rules, an incremental reconciliation is
added for the rules to skip kernel connection tracking.

This is needed to speed up reconciliation related to a change to no
track rules only.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Wait for reconciliation for no track pods and proxy updates

The current manager implementation relies on the reconciler to rewrite
the rules in case of no track pods or proxy updates.  Unlike the
previous implementation, this is carried out asynchronously.

Since all the tests in the CI expects these operations to be executed
synchronously, a notification channel is returned and closed when the
reconciliation has been completed successfully.

Despite this should not be strictly needed, waiting for the iptables
rules to be updated avoids a lot of flakes in the current e2e tests
implementation.

Being needed just for backward compatibility with the tests suite, it
may be removed in the future when the tests are rewritten to be less
unforgiven.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add a test for the reconciler

Add a unit-style integration test for the reconciler. The test verifies
that the reconciler emits the correct desired state in a handful of
scenarios (devices updates, local node allocation CIDR updates, proxy
rules and so on).

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

lmb commented on pkg/datapath/types/loader.go:
```diff
@@ -57,7 +57,7 @@ type Proxy interface {
 	// InstallProxyRules creates the necessary datapath config (e.g., iptables
 	// rules for redirecting host proxy traffic on a specific ProxyPort)
-	InstallProxyRules(proxyPort uint16, isLocalOnly bool, name string)
+	InstallProxyRules(proxyPort uint16, localOnly bool, name string) <-chan struct{}
```
> Being needed just for backward compatibility with the tests suite, it
may be removed in the future when the tests are rewritten to be less
unforgiven.

:sweat_smile: 

So at runtime we need this to be async, while during testing it should be sync? Looking at the code paths in endpoint it seems like those also want to be sync? My gut feeling would be to just keep the old interface and hide the channel read.

If you decide to stick with this: Nit: document when the returned channel is closed?

pippolo84 replied:
> So at runtime we need this to be async, while during testing it should be sync?

That's unfortunately the case. :cry: 
I've tried many other solutions to avoid performing synchronous operations but with no luck (see the CI runs in https://github.com/cilium/cilium/pull/31024, https://github.com/cilium/cilium/pull/30918, https://github.com/cilium/cilium/pull/30853 and https://github.com/cilium/cilium/pull/30821).
I've also failed to identify a reasonably small set of tests to update, that's why in the end I decided to go with this and finalize the PR. This refactor is also blocking the runtime devices reconciliation effort https://github.com/cilium/cilium/pull/30024

> Looking at the code paths in endpoint it seems like those also want to be sync? My gut feeling would be to just keep the old interface and hide the channel read.

On the one hand exposing the channel makes it clearer that the operation can (and should) be performed asynchronously, but on the other hand it is true that all the iptables consumers are written for the previous synchronous version.
Let's hide the channel as you suggested, it is probably better to keep it cleaner until we refactor the consumers (and the tests). :+1:

pippolo84 committed:
iptables: Do not block in case of a failed reconciliation

In case of a failed iptables reconciliation, unblock the waiting
goroutines closing the channel. This is done to avoid keeping the
consumer goroutines stuck forever and potentially blocking other and
more important agent operations.

In case of an error, the reconciler retries the iptables rules update at
a later time, in order to achieve eventual consistency.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

tommyp1ckles commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+func (lni localNodeInfo) equal(other localNodeInfo) bool {
+	if !lni.internalIPv4.Equal(other.internalIPv4) ||
```
nit: would this be easier to read if this was in the form:

```go
if lni.internalIPv4.Equal(other.internalIPv4) && lni.internalIPv6.Equal(other.internalIPv6) && ...
```

tommyp1ckles commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,203 @@
+			}
+
+			if err := updateRules(state, firstInit); err != nil {
+				log.WithError(err).Warning("iptables rules full reconciliation failed, will retry a full reconciliation")
```
nit: repeated instances of this seem serious, should this just be an error?

Furthermore, if there was some kind of problem this could begin emit a lot of similar logs (i.e. 5 a second).

pippolo84 replied:
Turned into an error and used a logging rate limiter.

tommyp1ckles commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1699,6 +1691,46 @@ func (m *Manager) addCiliumNoTrackXfrmRules() (err error) {
+func (m *Manager) removeNoTrackRules(addr netip.Addr, port uint16) error {
+	// Do not remove per endpoint NOTRACK rules if we are already skipping
+	// conntrack for all pod traffic.
+	if addr.Is4() && skipPodTrafficConntrack(false, m.sharedCfg.InstallNoConntrackIptRules) ||
```
nit: this is repeated, we could turn this into a helper func.

tommyp1ckles commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,203 @@
+			}
+			state.devices = newDevices
+			stateChanged = true
+		case localNode, ok := <-localNodeEvents:
```
There's a couple cases where closing one of the channels stops the loop (as well as context being closed).  If this were to happen it may be confusing as to what the reason for stopping the loop was.

pippolo84 replied:
`localNodeEvents` will close when the reconciler `ctx` closes. That `ctx` is the one from the job, so it will be closed when the job group stop hook is called. All the other channels `params.proxies`, `params.addNoTrackPod` and `params.delNoTrackPod` are closed when the `manager.Stop()` is called, that is, when the hive is stopped.
IOW, we can break out of the loop for that reason only.

tommyp1ckles commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,295 @@
+			if !firstInit {
+				// first init not yet completed, proxy rules will be updated as part of that
+				stateChanged = true
+				updatedChs = append(updatedChs, req.updated)
```
Should we check if the request updated channel is nil? or do we expect every request to have a wait channel.

pippolo84 replied:
I think it is safe to avoid checking, since the usage of the notification channel is an internal implementation detail in the iptables package. Also, we expect to have a valid channel for each request, since the channels are created and closed in the manager lifecycle hooks, and the reconciler goroutine is managed by a job appended to the same lifecycle.

tommyp1ckles commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -143,15 +144,32 @@ stop:
+				continue
+			}
+
+			if err := updateProxyRules(newProxyInfo.port, newProxyInfo.isLocalOnly, newProxyInfo.name); err != nil {
```
I'm not too familiar with how we do iptables reconciliation currently, but previous commits mention that the ordering is important here.  Is it ok to have proxy specific rules be installed on their own?

More broadly, what part of the ordering of these rules does this implementation maintain that the generic reconciler can't do?

pippolo84 replied:
> I'm not too familiar with how we do iptables reconciliation currently, but previous commits mention that the ordering is important here. Is it ok to have proxy specific rules be installed on their own?

It's safe to install proxy rules on their own, as long as we reinstall them when we do a "full reconciliation" (see `doInstallRules`). In that case, we rename all the chains with an `OLD_` prefix and we reinstall everything from scratch, so we need to reinstall the still valid proxy rules again.

> More broadly, what part of the ordering of these rules does this implementation maintain that the generic reconciler can't do?

The generic reconciler only works on statedb tables, but the statedb tables have no concept of ordering between entries. Therefore, using a statedb table to store iptables rules would have lost information regarding the order in which the rules should be appended to each chain.
Since we cannot use a statedb table, generic reconciler framework was ruled out, that's why iptables relies on a custom reconciler.

tommyp1ckles replied:
ok, thanks for the explanation.

tommyp1ckles commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -143,15 +144,32 @@ stop:
+			}
+
+			if err := updateProxyRules(newProxyInfo.port, newProxyInfo.isLocalOnly, newProxyInfo.name); err != nil {
+				log.WithError(err).Warning("iptables proxy rules incremental update failed, will retry a full reconciliation")
```
Same issue as in the ticker.C batch case, this could cause a lot of sudden error logs.

pippolo84 replied:
I've added rate limiting for errors logging in both full and partial reconciliation paths. I used two distinct rate limiters so that the partial errors (potentially very frequent) won't interfere with the more important full reconciliation error logs.

tommyp1ckles reviewed (commented):
Nice work! - Couple comments/questions.

Also, this may be a bit deeper on sig-proxy specific stuff than I can give a review, we may want to ping someone with a bit more experience here.

tommyp1ckles commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,295 @@
+				stateChanged = true
+				updatedChs = append(updatedChs, req.updated)
+			} else {
+				health.OK("iptables proxy rules incremental update completed")
```
Just some thoughts about this,  up to now most of the reconciliation tasks we've added health checking on are relatively infrequent (i.e. such as on a 60 second cron).

Theoretically, this health update has no upper bound on how much it can possibly be updating in which case this could be flapping between degraded/ok quickly (or even in the stable case, this would create a lot of statedb update events as updates that don't change health state still have a counter and timestamp updated). 

Furthermore, if the proxy update succeeds, then this only means those specific updates went through, which may overwrite a rightfully degraded status from a previous "general" reconciliation attempt. 

We may want to just let the ticker.C case handle all the (loop) health updates, to avoid these issues.

pippolo84 replied:
> Furthermore, if the proxy update succeeds, then this only means those specific updates went through, which may overwrite a rightfully degraded status from a previous "general" reconciliation attempt.

That's right, nice catch!

> We may want to just let the ticker.C case handle all the (loop) health updates, to avoid these issues.

Yep, that sounds a nice idea. I'll leave the log in case of a failed partial reconciliation attempt, and use health module only for the full reconciliation. :+1:

pippolo84 committed:
iptables: Simplify the helper for notrack rules support

Rework the helper to check if we should install notrack rules.

Notrack rules installation can be skipped in case of IPv6 or if
install-no-conntrack-iptables-rules option is set to true (that is, if
we are already skipping CT for all pod traffic).

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Rate limit errors logging

Add rate limiting to errors logging in both partial (proxy and no track
rules) and full reconciliation operations.

Two separate rate limiters are used so that they will not interfere with
each other.

Also, raise the log level to error in case of a failed reconciliaton.

Finally, improve the wording of the full reconciliation error log.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

tommyp1ckles reviewed (approved):
Looks good on my end!

jrajahalme commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -256,12 +256,15 @@ stop:
+			// do this even in case of a failed reconciliation, to avoid
+			// blocking consumer goroutines indefinitely.
+			for _, ch := range updatedChs {
+				close(ch)
```
Would it be better for the waiters to get the error value in the channel in case reconciliation failed?

pippolo84 replied:
With the current model I'm not sure that would be much useful: the reconciler keeps retrying the operations until success, so the information in the channel would describe the result of the first attempt only. The consumer won't have a chance to know what happens later (either persistent failure or successful recovery).
But yeah, we might change this in the future returning a channel that pumps each reconciling error and closes only after a success, if needed.

jrajahalme reviewed (approved):
One minor question for future enhancement :-)
